packages/utils/babylon-walk/src/scope.js

```diff
@@ -0,0 +1,219 @@
+      return this.parent.getRoot();
+    }
+
+    return this;
+  }
+
+  addBinding(name: string, decl: Node) {
+    this.names.add(name);
+    this.bindings.set(name, decl);
+  }
+
+  getBinding(name: string): ?Node {
+    return this.bindings.get(name) ?? this.parent?.getBinding(name);
+  }
+
+  addReference(identifier: Identifier) {
+    let references = this.references.get(identifier.name);
+    if (!references) {
+      references = new Set();
+      this.references.set(identifier.name, references);
+    }
+
+    references.add(identifier);
+  }
+
+  rename(from: string, to: string) {
+    // If already renamed, update the original to the final name.
+    let renamed = this.inverseRenames.get(from);
+    if (renamed) {
+      this.renames.set(renamed, to);
+    } else {
+      this.renames.set(from, to);
+      this.inverseRenames.set(to, from);
+    }
+  }
+
+  exit() {
+    // Rename declarations in this scope.
+    for (let [from, to] of this.renames) {
+      if (!this.names.has(from) && this.parent) {
+        this.parent.rename(from, to);
+      }
+
+      let references = this.references.get(from);
+      if (!references) {
+        continue;
+      }
+
+      for (let id of references) {
+        id.name = to;
+      }
+    }
+
+    // Propagate unknown references to the parent scope.
+    let parent = this.parent;
+    if (parent) {
+      for (let [name, ids] of this.references) {
+        if (!this.names.has(name)) {
+          for (let id of ids) {
+            parent.addReference(id);
+          }
+        }
+      }
+    }
+  }
+}
+
+export type ScopeState = {|
+  scope: Scope,
+|};
+
+export let scopeVisitor: Visitors<ScopeState> = {
+  Program(node, state) {
+    state.scope = new Scope('program');
+  },
+  Scopable: {
+    enter(node, state, ancestors) {
+      if (
+        !t.isScope(node, ancestors[ancestors.length - 2]) ||
+        t.isProgram(node) ||
+        t.isFunction(node)
+      ) {
+        return;
+      }
+      state.scope = new Scope('block', state.scope);
+    },
+    exit(node, state, ancestors) {
+      if (
+        !t.isScope(node, ancestors[ancestors.length - 2]) ||
+        t.isProgram(node)
+      ) {
+        return;
+      }
+
+      state.scope.exit();
+      if (state.scope.parent) {
+        state.scope = state.scope.parent;
+      }
+    },
+  },
+  Declaration: {
+    exit(node, {scope}) {
+      if (t.isFunction(node) || t.isExportDeclaration(node)) {
+        return;
+      }
+
+      // Register declarations with the scope.
+      if (isVariableDeclaration(node)) {
+        for (let decl of node.declarations) {
+          let ids = t.getBindingIdentifiers(node);
+          for (let id in ids) {
+            scope.addBinding(id, decl);
+            scope.addReference(ids[id]);
+          }
+        }
+      } else {
+        let ids = t.getBindingIdentifiers(node);
+        for (let id in ids) {
+          scope.addBinding(id, node);
+          scope.addReference(ids[id]);
```

Comment:
So references are not just references but also the identifier in the declaration?

Reply:
Hmm yeah I think this came from my other PR which was doing renaming. Could probably get rid of it